From ee5e59f9196c8f66d6b66397bb77ac63b4a436dd Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Wed, 18 Jul 2018 19:04:50 -0500
Subject: [PATCH] PCI_INTERRUPT_PIN Should Always Read 0

For SRIOV virtual functions, PCI_INTERRUPT_PIN
should always be 0.  Some SRIOV devices don't
adhere to this...

Signed-off-by: John Jacques <john.jacques@intel.com>
---
 drivers/vfio/pci/vfio_pci.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/vfio/pci/vfio_pci.c b/drivers/vfio/pci/vfio_pci.c
index 695b9d1..1afbca5 100644
--- a/drivers/vfio/pci/vfio_pci.c
+++ b/drivers/vfio/pci/vfio_pci.c
@@ -424,6 +424,16 @@ static int vfio_pci_get_irq_count(struct vfio_pci_device *vdev, int irq_type)
 {
 	if (irq_type == VFIO_PCI_INTX_IRQ_INDEX) {
 		u8 pin;
+
+		/*
+		 * VF's of SRIOV devices must have intx as 0.
+		 * Some devices might have this set due due to device bugs.
+		 * We could clear them since they aren't architecturally
+		 * allowed
+		 */
+		if (vdev->pdev->is_virtfn)
+			return 0;
+
 		pci_read_config_byte(vdev->pdev, PCI_INTERRUPT_PIN, &pin);
 		if (IS_ENABLED(CONFIG_VFIO_PCI_INTX) && !vdev->nointx && pin)
 			return 1;
-- 
2.7.4

